### DISCUSSION TITLE
How to don't wait for new page to load when wait and intercept an API response

### DISCUSSION THREAD

**Discussion by szanelato** [cypress-io/cypress#15849] [Questions and Help]
I'm facing with a behavior that's messing my assertions and my tests is failing.

I'm intercepting a API call and gave him a alias so at some point in my code I'm waiting for this alias and intercepting the POST call to validate the http status code and a value from a specific path inside the response body.

**The problem is:** I don't ask to wait for a new page to load and Cypress is waiting for that. In my CI environment the page that I'm getting the redirect in this scenario is too slow, so I do this by intercept the request to validate the response instead wait for redirect and assert the redirection url.

**The question is:**  how can I verify what I want inside a API response without wait for a new page to load?

I decided to use Cypress instead Selenium because of the Cypress cy.wait() and cy.intercept() features for verify responses from API calls instead to wait a page to load. I'm desapointed to be stucked in the same problem that I have using Selenium.

My code:

![image](https://user-images.githubusercontent.com/59705574/113880360-cec4fb00-9791-11eb-9aae-c3cb014e884d.png)
![image](https://user-images.githubusercontent.com/59705574/113880158-9d4c2f80-9791-11eb-8936-8cea7c864b31.png)
![image](https://user-images.githubusercontent.com/59705574/113880790-2ebba180-9792-11eb-9ab6-e3fa43673e6a.png)

**Comment by bahmutov** [comment]
So you are trying to confirm the page loads BUT do not want to wait for the entire body of the page to load? The server is responding with the page, that's the slow part, not the load after that in the browser.

**Reply by szanelato** [reply]
@bahmutov Actually I would like to validate the test by my response body verification that I'm performing in my service and ignore the page load.

**Reply by bahmutov** [reply]
So in that case, you probably do not want to visit this page and instead use cy.request to grab it as HTML

**Reply by szanelato** [reply]
@bahmutov I am testing a login app via browser and my back-end login service returns me the destination url for the user to be redirected after login by my web app. In my code I intercept the response from my service to meet the condition of my test and I would like to eliminate the need to wait for a new page (which is the one returned by the service) to be loaded. Here is the question number 1: It's that possible?

```
verifyLoginSuccess () {
    cy.intercept('https://mylogin.service.com/authenticate').as('postAuthenticate');
    cy.wait ('@ postAuthenticate'). then (interception => {
         expect (interception.response.statusCode) .equal (200);
         expect (interception.response.body.dest) .equal ('https://destinationurl.com');
     })
} 
```

So with this function I was expecting that my tests get passed after validate the statusCode and the specific property inside response body throught the cy.intercept() function, but cypress is forcing me to wait for new page to load.

I'm sorry if I wasn't clear before.

**Reply by iivok** [reply]
@bahmutov 
I'm running into this same issue. I'm testing a login page with Cypress 12 and am trying to create a test that just goes through the login.

- There's a form that POSTs on submit (I click the button in the test)
- The response from the POST is a redirect to another url (GET)
- A loading page is shown for a bit and another POST call _authenticate_ is sent via form submit
- The _authenticate_ response contains a redirect to the actual page

I have set interceptors and cy.waits with assertions for each of the calls and it all works as expected when there's no errors, which there usually aren't when I run that test alone.

However, when I run that test after another test, I get a 403 from the _authenticate_ call like 90% of the time and when that happens, Cypress hangs with "waiting for the page to load". The waits don't finish and I can't seem to get Cypress to just fail instead of waiting for an eventual timeout.

Edit: 
I managed to fail the test immediately with something like this:
```
cy.intercept('POST', '**/authenticate/', (request) => {
        request.on('before:response', (res) => {
          if (res.statusCode === 403) {
            throw new Error('403 response from authentication');
          }
        });
        request.continue();
      })
```

Also using fetch with error handling instead of form submit on the auth loader page helps with the unfinished page load.

**Comment by handipriyono** [comment]
hi @szanelato szanelato, have you find the solution ??
